### PR TITLE
Added SQLiteKeyInfoManager Storage & Retrieval Functionality.

### LINKED ISSUE #503: Add Basic SQLiteKeyInfoManager Storage/Retrieval Functionality
Add the `SQLiteKeyInfoManager`, implementing the `ManageKeyInfo` trait & methods.

### PR DESCRIPTION
**Pull request by MattDavis00**
This PR adds basic storage and retrieval mechanisms into the `SQLiteKeyInfoManager`.
These mechanism satisfy our current unit tests which have been copied from the `OnDiskKeyInfoManager` for now.
Additional unit tests need adding to affirm the `key_name`, `application_name`, `authenticator` namespace we defined during design.
Furthermore, the `kim_metadata` table does not currently exist and will need adding.

(Won't be perfect, just doing it bit by bit for ease of PRs sake)

Closes #503

Signed-off-by: Matt Davis <matt.davis@arm.com>

### CODE REVIEW

File: src/key_info_managers/sqlite_manager/mod.rs [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+        let key_info_blob = bincode::serialize(&key_info).map_err(|e| {
+            format_error!("Error serializing key info", e);
+            RusqliteError::ToSqlConversionFailure(e)
+        })?;
+
+        // Insert the new key mapping, if a record does not exist.
+        // If one does exist, update the existing record.
+        let _ = conn.execute(
+            "
+            INSERT OR IGNORE INTO
```
Comment (ionut-arm) [inline src/key_info_managers/sqlite_manager/mod.rs:159]: Does the `INSERT OR IGNORE` mean that if one of the insertions fails, it fails silently (i.e., no error is returned by the KIM)?

File: src/key_info_managers/sqlite_manager/mod.rs [hunk 2]
```diff
@@ -0,0 +1,597 @@
...
+                        let error = Box::new(Error::new(ErrorKind::InvalidData, e));
+                        RusqliteError::FromSqlConversionFailure(64, Integer, error)
+                    })?,
+                ),
+                ProviderIdentity::new(row.get("provider_uuid")?, row.get("provider_name")?),
+                row.get("key_name")?,
+            );
+            let key_info_blob: Vec<u8> = row.get("key_info")?;
+
+            // TODO: Change this to (protobuf?) version once format has been decided.
```
[the remark this answers is not in the document]
Comment (MattDavis00) [reply src/key_info_managers/sqlite_manager/mod.rs:109]: Would probably be under the TPM discussions here? https://github.com/parallaxsecond/parsec/issues/504

[the remark this answers is not in the document]
Comment (ionut-arm) [reply src/key_info_managers/sqlite_manager/mod.rs:109]: Hmm, I think it should a separate issue that depends on #504 , as they can be done independently.

File: src/key_info_managers/sqlite_manager/mod.rs [hunk 3]
```diff
@@ -0,0 +1,597 @@
...
+            RusqliteError::ToSqlConversionFailure(e)
+        })?;
+
+        // Insert the new key mapping, if a record does not exist.
+        // If one does exist, update the existing record.
+        let _ = conn.execute(
+            "
+            INSERT OR IGNORE INTO
+                `key_mapping`
+                (`authenticator_id`, `application_name`, `provider_uuid`, `provider_name`, `key_name`, `key_info`, `key_info_version`)
```
[the remark this answers is not in the document]
Comment (MattDavis00) [reply src/key_info_managers/sqlite_manager/mod.rs:161]: My current plan is before these queries are run, check whether any keys need to be migrated to the current version. If they do need upgrading; either throw an error if we end up deciding on a standalone migration program (e.g. "please use the key info migration program to upgrade your keys"), or automatically upgrade the keys within the parsec service if this is the approach we decide. If all keys are up to date it will reach this logic and load the KIM mappings.

TLDR; should never hit these load statements with an incorrect `key_info_version`.

File: src/key_info_managers/sqlite_manager/mod.rs (hunk 1 above)
Comment (MattDavis00) [reply src/key_info_managers/sqlite_manager/mod.rs:159]: The `INSERT OR IGNORE` clause will attempt to `INSERT` first; if there is already a record that exists then this will fail (silently, not throwing a `RusqliteError`). For every request it attempts to `UPDATE`. If the `UPDATE` fails, this will not be silent and will generate a `RusqliteError`.

Essentially it's "making sure the record exists", and then "updating it to the correct value".

File: src/key_info_managers/sqlite_manager/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (MattDavis00) [reply src/key_info_managers/sqlite_manager/mod.rs:109]: I'll create a new issue, something along the lines of `Decide upon a new serialization format for KeyInfo` sound good?

[the remark this answers is not in the document]
Comment (ionut-arm) [reply src/key_info_managers/sqlite_manager/mod.rs:109]: I think the question should actually be - are you planning to implement that in this PR? Or to leave it for a subsequent one? If you're just going to wait to implement it here then there's no need for a new issue. Otherwise something like `Decide and implement...` works

File: src/key_info_managers/sqlite_manager/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ionut-arm) [reply src/key_info_managers/sqlite_manager/mod.rs:161]: Cool, thanks for the info! 👍🏻

File: src/key_info_managers/sqlite_manager/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (MattDavis00) [reply src/key_info_managers/sqlite_manager/mod.rs:109]: Planning on leaving it to a subsequent PR. I'll create an issue.

File: src/key_info_managers/sqlite_manager/mod.rs (hunk 1 above)
Comment (ionut-arm) [reply src/key_info_managers/sqlite_manager/mod.rs:159]: Thanks for the explanation! 

This might sound strange, but would it make sense to split this into the two operations, `INSERT` and `UPDATE`, let the `INSERT` fail and "manually" ignore it if it does (by, for example, dropping the result from `.execute()`)? Don't know if this makes the whole thing less efficient or if it's just good practice for SQL queries to be done together like this, though

Comment (MattDavis00) [reply src/key_info_managers/sqlite_manager/mod.rs:159]: On further investigation there is a more semantically correct `REPLACE` operation.
https://dev.mysql.com/doc/refman/8.0/en/replace.html
This should result in one query such as something like:
```
REPLACE INTO
    `key_mapping`
    (`authenticator_id`, `application_name`, `provider_uuid`, `provider_name`, `key_name`, `key_info`, `key_info_version`)
VALUES
    (?1, ?2, ?3, ?4, ?5, ?6, ?7);
```
If the record does not exist, it inserts it. If it does, then it deletes the existing record and inserts the new one. (This occurs by matching on the primary/unique indexes which in our case are `authenticator_id`-`application_name`-`key_name`)

Comment (MattDavis00) [reply src/key_info_managers/sqlite_manager/mod.rs:159]: > I'll have a look into implementing this and whether it is supported in SQLite.

Update: seems to work fine. Will push a revised PR commit.

**Review by ionut-arm — APPROVED** [review]
Looks good, thanks!
